### ISSUE TITLE
Question about the input shape in transforms.scale.py

### ISSUE THREAD

**Issue by jialee93** [vita-epfl/openpifpaf#183]
Hi @svenkreiss. I have a doubt about the input image shape. As I know, you probably resize the length of the input image to an odd number such as n*stride+1 on the assumption that the pixels are dots without areas.  The relevant code is here:

https://github.com/vita-epfl/openpifpaf/blob/21baabf9c6bbd0bea3e8e465a726abfa8dbeeccf/openpifpaf/transforms/scale.py#L25

But why didn't you resize the smaller side of the image to an odd number? For instance, you resize an image to (641, 480) instead of (641, 481). Will this operation brings about precision loss?

**Comment by svenkreiss** [comment]
I am not changing the aspect ratio. The other axis is what it is.

**Comment by jialee93** [comment]
Thank you for your reply😊. But what if we pad the image first to have odd width and odd height and then scale the image. I think such operation sounds more reasonable. And in this way the aspect ratio of the original image is not changed. Please correct me if I'm wrong.

**Comment by svenkreiss** [comment]
Depending on what you are doing, but for training and for evaluation with
batch size > 1, padding happens after rescaling.
You are right, for batchsize=1 during evaluation a padding might sometimes
be helpful, but I have observed no significant change in the coco metric
between evaluations with batchsize 1 or larger 1.

**Comment by jialee93** [comment]
Thank you @svenkreiss, I ran your code with different scale operations:

https://github.com/vita-epfl/openpifpaf/blob/21baabf9c6bbd0bea3e8e465a726abfa8dbeeccf/openpifpaf/transforms/scale.py#L27
https://github.com/vita-epfl/openpifpaf/blob/238d89ee1d26d03e397aae9c66493f7085d71fb7/openpifpaf/transforms.py#L69

The COCO metrics are indeed nearly the same :)

**Comment by jialee93** [comment]
Hi @svenkreiss, sorry for bothering you again. As you pointed out here:
https://github.com/vita-epfl/openpifpaf/blob/51b6d80951460d6aad51cd7ab09e86bfeec1c667/openpifpaf/transforms/__init__.py#L6
the center of the first pixel (left-top) in the image numpy array actually lies at (0.5, 0,5) in the Cartesian pixel coordinate system. As for the COCO float keypoint and bbox annotations w.r.t the origin of the coordinate system, I think the origin point (0, 0) also lies at the upper left corner of the image numpy array. But I cannot find reliable reference about my hypothesis.

If my hypothesis is true, the image float length is just the same as the image pixel length (corresponds to the red length in the following picture). 
![](https://raw.githubusercontent.com/jialee93/PictureBed/master/img2bolg/20200224105449.png)
If we resize the original image two times using PIL package, it uses geometric center alignment just as OpenCV cv2.resize() function (regard the pixel as a cell with 1*1 area). Then the image transformation mismatches the float keypoint coordinates transformation in your code. 

If we transform the keypoint coordinates using:

https://github.com/vita-epfl/openpifpaf/blob/238d89ee1d26d03e397aae9c66493f7085d71fb7/openpifpaf/transforms.py#L69

then, the image and keypoint transformations matches, but the image length calculation changes such that the coordinate x becomes w-x during flip transformation which is illustrated as follows (I hope my illustration is clear):
![](https://raw.githubusercontent.com/jialee93/PictureBed/master/img2bolg/20200224111447.jpeg)

Even if the influence is not significant, I just would like to make things clear. Hope you could give me a help. Thank you!

**Comment by svenkreiss** [comment]
Thanks for your detailed report. All coordinates in PifPaf have (0, 0) at the center of the top-left pixel, not at the top-left corner of the top-left pixel. Given the padding conventions in the PyTorch models, this means that (0, 0) is also at the center of the top-left feature map cell.

Does this clarify things?

(I am happy you are checking this. Please let me know if you find some inconsistency. It could be there is something wrong somewhere. Thanks!)

**Comment by jialee93** [comment]
First, we do not know the normalization of COCO source image and keypoint annotation standard. But this does not matter if we keep our image transformation and keypoint coordinate transformation consistent. The consistent decoding can make up for the encoding shifting.

Second, the PIL image.resize() actually resize the image with geometric center alignment, which means the center of the top-left pixel as (0.5, 0.5). For instance, we resize the 1-D array two times 
 [0,1,2,3,4,5] --> resize with  geometric center alignment--> [-0.25, 0.25, 0.75, 1.25, 1.75, 2.25, 2.75, 3.25, 3.75, 4.25, 4.75, 5.25] 
--> (reflection border mode) --> [0, 0.25, 0.75, 1.25, 1.75, 2.25, 2.75, 3.25, 3.75, 4.25, 4.75, 5].

Thus, the image transformation mismatched the keypoint coordinate transformation in the current PIFPAF.

Thank you @svenkreiss, I'm glad you're interested in what I'm thinking.  Thanks to your discussion.

**Comment by svenkreiss** [comment]
Thanks for clarifying. Your example is very clear and helpful.
Do you have a reference for the PIL image resize coordinates?
Or can you share the code for your above resizing test with PIL?

If you want, feel free to submit this as a unit test in a Pull Request,
e.g. `test_pil_geometric_center_alignment` so that we know for future
versions of PIL as well that this is what it does.

**Comment by jialee93** [comment]
If the COCO keypoint coordinates are annotated w.r.t the most left side of the image edge, then the flip operation should be: x' -> w - x. 
Otherwise (w.r.t x=0.5), the flip operation should be: x' -> w - 1 - x. In this case , the version 9 of PIFPAF is just as what I described above. 
Perhaps, the PIFPAF's strict offset encoding could alleviate what we concern. I haven't browsed much your pretty code yet. I'll let you know if I figured out these.  Again, thank you for your instructions and patience.

**Comment by svenkreiss** [comment]
Thanks.
If the PIL image resize really is not consistent, we will fix it there.
Otherwise, the coordinate system used internally in pifpaf needs to stay as
is: (0,0) is the center of the top left pixel. (And also the center of the
top-left feature map cell which is the more important thing.)
If the COCO annotations are not like that, we will convert them by
subtracting 0.5 when reading them in (might already happen).

So we are good with the flip operation then?
Please keep us posted here in this issue about what you find out.

**Comment by jialee93** [comment]
I had already corrected my earlier comment and I made some conclusions here:

The PIL/OpenCV  image resize still regard the center of the top-left image cell as the origin (0, 0) because the indices start from zero not 0.5, and the pixels are regarded to lie at the centers of image cells. In other words, the pixels have areas. I note this to clarify the terms we didn't  properly use earlier.

In OpenCV, cv2.warpaffine() method treats the pixels as dots without areas. In the current PIFPAF (version 0.10), the keypoint coordinate transformation also hold this perspective. However, PIL image resize not.

In PIFPAF version 0.9 or earlier, the image transformations (exclude the flip, we are not certain) keep consistent with the keypoint coordinate transformations.

The only remaining issue you and me haven't made clear is that the COCO keypoint and bbox annotation standards. We do not know the coordinate reference: the most left side of the image border, or the vertical line crossing the centers of the first column image cells.

**Comment by jialee93** [comment]
Supplement :)

I have browsed all the image and keypoint coordinate transformations, and I found that the rotate transformation only considers 0, 90, 180, 270 rotations. Dose rotating 270 degrees make sense? I have never tried such extent which means upside down.

Other float degrees are not performed during training. The image rotation and keypoint coordinate rotation should also keep consistent if float degrees are employed (but only 0, 90, 180, 270 rotations are executed in the code):
https://github.com/vita-epfl/openpifpaf/blob/51b6d80951460d6aad51cd7ab09e86bfeec1c667/openpifpaf/transforms/rotate.py#L53
